### ISSUE TITLE
With a GPU, memory continues to grow without releasing

### ISSUE THREAD

**Issue by longzhendong** [eclipse/deeplearning4j#8236]
#### Version Information

Please indicate relevant versions, including, if relevant:

* Deeplearning4j 1.0.0-beta5
*Windows10
* CUDA 9.2

1、Network structure is as follows:
Nd4j.getMemoryManager().setAutoGcWindow(3000);

		ComputationGraphConfiguration config = new NeuralNetConfiguration.Builder()
				.weightInit(WeightInit.RELU).activation(Activation.RELU).updater(new Adam(0.01))
				.l2(0.0001).graphBuilder().addInputs("input")
				.setInputTypes(InputType.recurrent(dict.size()))
				.addLayer("embedding",
						new EmbeddingLayer.Builder().nIn(dict.size()).nOut(100).build(), "input")
				.addLayer("selfAttention1",
						 new SelfAttentionLayer.Builder().headSize(30).nHeads(4)
							.nIn(100).nOut(100)
							.projectInput(true).build(),
						"embedding")
				.addLayer("selfAttention2",
						 new SelfAttentionLayer.Builder().headSize(30).nHeads(4)
							.nIn(100).nOut(100)
							.projectInput(true).build(),
						"selfAttention1")
				.addLayer("selfAttention3",
						 new SelfAttentionLayer.Builder().headSize(30).nHeads(4)
							.nIn(100).nOut(100)
							.projectInput(true).build(),
						"selfAttention2")
				.addLayer("globalPool",
						new GlobalPoolingLayer.Builder().poolingType( PoolingType.AVG).build(), "selfAttention3")
				.addLayer("out1",
						new OutputLayer.Builder().lossFunction(LossFunctions.LossFunction.MCXENT)
								.activation(Activation.SOFTMAX).nIn(100).nOut(2) 
								.build(),
						"globalPool")
				.setOutputs("out1").build();

		ComputationGraph net = new ComputationGraph(config);

2、The JVM parameter
-Xms1G -Xmx3G -Dorg.bytedeco.javacpp.maxbytes=5G -Dorg.bytedeco.javacpp.maxphysicalbytes=8G -Dorg.deeplearning4j.ui.port=9001

3、Memory keeps growing

![image](https://user-images.githubusercontent.com/30429640/64839376-e2e25e00-d629-11e9-9ac9-398a27e6dc89.png)

4、The exception stack information is as follows

Exception in thread "main" java.lang.RuntimeException: Op [multi_head_dot_product_attention_bp] execution failed
	at org.nd4j.linalg.jcublas.ops.executioner.CudaExecutioner.exec(CudaExecutioner.java:2307)
	at org.nd4j.autodiff.samediff.internal.InferenceSession.getOutputsHelper(InferenceSession.java:505)
	at org.nd4j.autodiff.samediff.internal.InferenceSession.getOutputs(InferenceSession.java:119)
	at org.nd4j.autodiff.samediff.internal.InferenceSession.getOutputs(InferenceSession.java:56)
	at org.nd4j.autodiff.samediff.internal.AbstractSession.output(AbstractSession.java:335)
	at org.nd4j.autodiff.samediff.SameDiff.directExecHelper(SameDiff.java:3181)
	at org.nd4j.autodiff.samediff.SameDiff.execBackwards(SameDiff.java:4693)
	at org.nd4j.autodiff.samediff.SameDiff.execBackwards(SameDiff.java:4657)
	at org.nd4j.autodiff.samediff.SameDiff.execBackwards(SameDiff.java:4666)
	at org.deeplearning4j.nn.layers.samediff.SameDiffLayer.backpropGradient(SameDiffLayer.java:159)
	at org.deeplearning4j.nn.graph.vertex.impl.LayerVertex.doBackward(LayerVertex.java:149)
	at org.deeplearning4j.nn.graph.ComputationGraph.calcBackpropGradients(ComputationGraph.java:2712)
	at org.deeplearning4j.nn.graph.ComputationGraph.computeGradientAndScore(ComputationGraph.java:1381)
	at org.deeplearning4j.nn.graph.ComputationGraph.computeGradientAndScore(ComputationGraph.java:1341)
	at org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:170)
	at org.deeplearning4j.optimize.solvers.StochasticGradientDescent.optimize(StochasticGradientDescent.java:63)
	at org.deeplearning4j.optimize.Solver.optimize(Solver.java:52)
	at org.deeplearning4j.nn.graph.ComputationGraph.fitHelper(ComputationGraph.java:1165)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:1115)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:1082)
	at org.deeplearning4j.AttentionClassify.main(AttentionClassify.java:155)
Caused by: java.lang.RuntimeException: [DEVICE] allocation failed; Error code: [2]
	at org.nd4j.linalg.jcublas.ops.executioner.CudaExecutioner.exec(CudaExecutioner.java:2491)
	at org.nd4j.linalg.jcublas.ops.executioner.CudaExecutioner.exec(CudaExecutioner.java:2297)
	... 20 more

**Comment by AlexDBlack** [comment]
I've attempted to reproduce this using the following code: https://gist.github.com/AlexDBlack/2b1c90caf9d6768ab80a6ae9ec835966

On Windows 10, CUDA 10.1, 1.0.0-beta5, Titan X.

Run for around 30 minutes without UI, and then separately 30 minutes with UI. GPU memory use is fine/flat in both cases.

@longzhendong Can you reproduce the problem with my code on your system? If not, maybe your iterator could be the cause?

**Comment by zdl010** [comment]
@AlexDBlack Can you upload it to github? I can't open the link

**Comment by AlexDBlack** [comment]
@longzhendong 
[no ui.txt](https://github.com/eclipse/deeplearning4j/files/3686371/no.ui.txt)
[with ui.txt](https://github.com/eclipse/deeplearning4j/files/3686372/with.ui.txt)

**Comment by zdl010** [comment]
@AlexDBlack 
[Temp.txt](https://github.com/eclipse/deeplearning4j/files/3689918/Temp.txt)

**Comment by zdl010** [comment]
@AlexDBlack o.n.l.f.Nd4jBackend - Loaded [JCublasBackend] backend
ND4J CUDA build version: 9.2.148
CUDA device 0: [GeForce GTX 1070 Ti]; cc: [6.1]; Total memory: [8589934592];
o.n.n.NativeOpsHolder - Number of threads used for OpenMP: 32
o.n.n.Nd4jBlas - Number of threads used for OpenMP BLAS: 0
o.n.l.a.o.e.DefaultOpExecutioner - Backend used: [CUDA]; OS: [Windows 10]
o.n.l.a.o.e.DefaultOpExecutioner - Cores: [12]; Memory: [2.7GB];
o.n.l.a.o.e.DefaultOpExecutioner - Blas vendor: [CUBLAS]
o.n.l.j.o.e.CudaExecutioner - Device Name: [GeForce GTX 1070 Ti]; CC: [6.1]; Total/free memory: [8589934592]
o.d.n.g.ComputationGraph - Starting ComputationGraph with WorkspaceModes set to [training: ENABLED; inference: ENABLED], cacheMode set to [NONE]
o.d.u.p.PlayUIServer - DL4J UI Server started at http://localhost:9001
o.d.u.p.PlayUIServer - StatsStorage instance attached to UI: InMemoryStatsStorage(uid=55df9f56)
o.d.o.l.ScoreIterationListener - Score at iteration 0 is 0.4422454710050051
o.d.o.l.ScoreIterationListener - Score at iteration 1 is 0.009122205386832165
o.d.o.l.ScoreIterationListener - Score at iteration 2 is 0.008160302039160173
o.d.o.l.ScoreIterationListener - Score at iteration 3 is 0.0072844484860167084
o.d.o.l.ScoreIterationListener - Score at iteration 4 is 0.006491813034736151
o.d.o.l.ScoreIterationListener - Score at iteration 5 is 0.005777965235731088
o.d.o.l.ScoreIterationListener - Score at iteration 6 is 0.005137845184374149
o.d.o.l.ScoreIterationListener - Score at iteration 7 is 0.004566140115974122
o.d.o.l.ScoreIterationListener - Score at iteration 8 is 0.004057437406258831
o.d.o.l.ScoreIterationListener - Score at iteration 9 is 0.0036063247514902088
o.d.o.l.ScoreIterationListener - Score at iteration 10 is 0.003207474896198781
o.d.o.l.ScoreIterationListener - Score at iteration 11 is 0.002855739415737631
Exception in thread "main" java.lang.RuntimeException: Op [multi_head_dot_product_attention_bp] execution failed
	at org.nd4j.linalg.jcublas.ops.executioner.CudaExecutioner.exec(CudaExecutioner.java:2307)
	at org.nd4j.autodiff.samediff.internal.InferenceSession.getOutputsHelper(InferenceSession.java:505)
	at org.nd4j.autodiff.samediff.internal.InferenceSession.getOutputs(InferenceSession.java:119)
	at org.nd4j.autodiff.samediff.internal.InferenceSession.getOutputs(InferenceSession.java:56)
	at org.nd4j.autodiff.samediff.internal.AbstractSession.output(AbstractSession.java:335)
	at org.nd4j.autodiff.samediff.SameDiff.directExecHelper(SameDiff.java:3181)
	at org.nd4j.autodiff.samediff.SameDiff.execBackwards(SameDiff.java:4693)
	at org.nd4j.autodiff.samediff.SameDiff.execBackwards(SameDiff.java:4657)
	at org.nd4j.autodiff.samediff.SameDiff.execBackwards(SameDiff.java:4666)
	at org.deeplearning4j.nn.layers.samediff.SameDiffLayer.backpropGradient(SameDiffLayer.java:159)
	at org.deeplearning4j.nn.graph.vertex.impl.LayerVertex.doBackward(LayerVertex.java:149)
	at org.deeplearning4j.nn.graph.ComputationGraph.calcBackpropGradients(ComputationGraph.java:2712)
	at org.deeplearning4j.nn.graph.ComputationGraph.computeGradientAndScore(ComputationGraph.java:1381)
	at org.deeplearning4j.nn.graph.ComputationGraph.computeGradientAndScore(ComputationGraph.java:1341)
	at org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:170)
	at org.deeplearning4j.optimize.solvers.StochasticGradientDescent.optimize(StochasticGradientDescent.java:63)
	at org.deeplearning4j.optimize.Solver.optimize(Solver.java:52)
	at org.deeplearning4j.nn.graph.ComputationGraph.fitHelper(ComputationGraph.java:1165)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:1115)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:1025)
	at org.deeplearning4j.Temp.main(Temp.java:79)
Caused by: java.lang.RuntimeException: [DEVICE] allocation failed; Error code: [2]
	at org.nd4j.linalg.jcublas.ops.executioner.CudaExecutioner.exec(CudaExecutioner.java:2491)
	at org.nd4j.linalg.jcublas.ops.executioner.CudaExecutioner.exec(CudaExecutioner.java:2297)
	... 20 more

**Comment by zdl010** [comment]
@AlexDBlack  JVM param :
-Xms1G -Xmx3G -Dorg.bytedeco.javacpp.maxbytes=7G -Dorg.bytedeco.javacpp.maxphysicalbytes=10G -Dorg.deeplearning4j.ui.port=9001

**Comment by zdl010** [comment]
![image](https://user-images.githubusercontent.com/30429640/66198379-dbfdb700-e6ce-11e9-801b-cdae241d412b.png)

**Comment by zdl010** [comment]
@AlexDBlack I uploaded a code, you try my code

**Comment by zdl010** [comment]
@AlexDBlack If you need to set the time step a little bit longer, you'll get a memory leak

**Comment by zdl010** [comment]
@AlexDBlack  This bug is the same problem

 https://github.com/eclipse/deeplearning4j/issues/8240

**Comment by AlexDBlack** [comment]
Thanks. I did run your code a couple of days ago, and (unlike my version) I was seeing unstable memory use - lots of allocations / deallocations. But no OOM even after an hour.
I'm running again now for longer with the same memory config (xms/xmx/javacpp) that you posted.

Separately, I'm working on improving memory use in SameDiff, which will likely impact this (the attention layers are implemented using SameDiff internally).

**Comment by AlexDBlack** [comment]
So, after running for around 5 hours - no evidence of a leak, but there is a lot out-of-workspace memory being allocated/deallocated.

Here's GPU memory use vs. time:

![image](https://user-images.githubusercontent.com/2360237/66466475-7e9ea700-eace-11e9-8255-7bc2cc64f34e.png)


I'll leave it running for a while longer. And I will re-test once my SameDiff improvements are done. I expect that will help a lot with this (and should also reduce peak memory requirements). Maybe I'm not seeing an OOM simply because I have a 12GB card (Titan X), and peak memory requirements greater than the 8GB that the 1070 Ti has.

**Comment by zdl010** [comment]
https://github.com/eclipse/deeplearning4j/issues/8240
Thanks！There is a zip in this link, you can test it

**Comment by AlexDBlack** [comment]
So, with the latest master (my SameDiff memory management PR was merged yesterday), memory consumption is definitely better:
![image](https://user-images.githubusercontent.com/2360237/67477945-d6303b80-f6a6-11e9-84bd-365ee3af1a46.png)

I still have a bit more work to do here, to reduce the number of allocations and deallocations.
And the min vs. max memory use here is still more than I would have expected for this architecture, so I want to look into it further before closing this.

Unfortunately snapshots aren't being updated at present, that will happen in the next few days hopefully.

**Comment by zdl010** [comment]
@AlexDBlack It's a great job！！

**Comment by AlexDBlack** [comment]
So, one more step: https://github.com/eclipse/deeplearning4j/issues/8329

Unfortunately this didn't help nearly as much as I expected here. There's definitely an improvement in memory use, but it's only a small improvement here. It'll help performance a bit though (eliminates a few memory copies).

![image](https://user-images.githubusercontent.com/2360237/68067248-4ff7b180-fd98-11e9-9bef-1ae1c5724b9c.png)

The fact that the memory use has not stabilized (i.e., still a big difference between minimum and maximum memory use over time) suggests it's probably the attention operations themselves that are at fault here... I can't see another explanation, and I'm still not happy with what I'm seeing here and will look into it further.

**Comment by AlexDBlack** [comment]
Also those improvements are now on eclipse master, they should be available on snapshots the next time they are built.
https://deeplearning4j.org/docs/latest/deeplearning4j-config-snapshots

**Comment by AlexDBlack** [comment]
As I previously expected: This is coming from the C++ attention ops.
Given the previous memory improvements already implemented/merged, I'll close this in favor of this issue: https://github.com/eclipse/deeplearning4j/issues/8345
